### PR TITLE
11 create crud operations for hotels

### LINKED ISSUE #11: Create CRUD operations for hotels
The urls must have the following format:
/api/hotel/1/ (read)
/api/hotel/1/create (create)
/api/hotel/1/update (update)
/api/hotel/1/delete (delete)

Note: When a hotel is created, a room must be created too by default.

### PR DESCRIPTION
**Pull request by georgepoul**
[Added a Hotel DTO, Hotel Service, Hotel Controller, modelMapper for Hotel that converts Hotel to HotelDTO, CRUD for Hotels and fixed same issues.

### CODE REVIEW

File: src/main/java/com/sphy/hotelmanagementapplication/controller/HotelController.java [hunk 1]
```diff
@@ -35,77 +32,93 @@ public class HotelController {
...
-//
-//        return  hotelDTO;
-  			return null;
+        return service.saveHotelDTO(hotelDTO);
     }
 
+    /***
+     * Create new hotels
+     * @param hotelsDTO is a list of hotels to be saved
+     * @return the list of hotels tha saved
```
Comment (DimK10) [inline src/main/java/com/sphy/hotelmanagementapplication/controller/HotelController.java]: that were saved (in comment)

File: src/main/java/com/sphy/hotelmanagementapplication/domain/Admin.java [hunk 2]
```diff
@@ -19,7 +19,10 @@ public Admin(Long id, boolean emailVerify, String username, String firstname, St
         super(id, emailVerify, username, firstname, lastname, email, Role.ADMIN);
     }
 
-	public Long getId() {
+    public Admin(long l) {
```
Comment (DimK10) [inline src/main/java/com/sphy/hotelmanagementapplication/domain/Admin.java]: use super.setId() to set the id for the newly created object that used this constructor

File: src/main/java/com/sphy/hotelmanagementapplication/dto/HotelDTO.java [hunk 3]
```diff
@@ -22,14 +20,15 @@ public class HotelDTO implements Serializable {
 
     private Set<RoomDTO> rooms =new HashSet<>();
 
-	public HotelDTO() {
-	}
+    public HotelDTO(long id) {
```
Comment (DimK10) [inline src/main/java/com/sphy/hotelmanagementapplication/dto/HotelDTO.java:25]: Same as Admin.java at line 22

File: src/main/java/com/sphy/hotelmanagementapplication/dto/RoomDTO.java [hunk 4]
```diff
@@ -24,29 +22,6 @@ public class RoomDTO implements Serializable {
 
 	private boolean disabled;
 
-	public RoomDTO(Long id, String name, Long hotel) {
-		Id = id;
-		this.name = name;
-		this.hotel = hotel;
-	}
-
-	public RoomDTO() {
```
Comment (DimK10) [inline src/main/java/com/sphy/hotelmanagementapplication/dto/RoomDTO.java:33]: If we are to sue the empty constructor, the best practice is to show it in class. Add the empty constructor please.

File: src/main/java/com/sphy/hotelmanagementapplication/service/HotelService.java [hunk 5]
```diff
@@ -105,35 +145,59 @@ public HotelDTO updateHotel(HotelDTO hotelDTO, List<RoomDTO> roomDTOS) throws Nu
 			existingHotel.setAreaName(hotelDTO.getAreaName());
 			Optional<Admin>  admin = adminRepository.findById(hotelDTO.getId());
 			admin.ifPresent(existingHotel::setOwner);
-			for (RoomDTO roomDTO : roomDTOS){
-				roomService.updateRoom(roomDTO);
-			}
+
 			hotelRepository.save(existingHotel);
```
Comment (DimK10) [inline src/main/java/com/sphy/hotelmanagementapplication/service/HotelService.java]: We need to return the modified entity here, the save method returns the object that was saved, if the save operation wa successful. Please get the Hotel object that was saved, convert it to DTO and return that, and not the initial HotelDTO

File: src/main/java/com/sphy/hotelmanagementapplication/service/HotelService.java [hunk 6]
```diff
@@ -105,35 +145,59 @@ public HotelDTO updateHotel(HotelDTO hotelDTO, List<RoomDTO> roomDTOS) throws Nu
...
+			if (hotelDTO.getRooms() == null){
+				throw new Exception("If you wont to save hotels you mast add one or more rooms");
+			}
+
 			hotels.add(hotelDTOToHotel.converter(hotelDTO));
 		}
 
-		roomService.saveRooms(roomsDTOS);
-
 		hotelRepository.saveAll(hotels);
```
Comment (DimK10) [inline src/main/java/com/sphy/hotelmanagementapplication/service/HotelService.java:201]: The saveAll returns an Iterable of the objects that were saved. Please return those after converting them to DTOS

File: src/main/java/com/sphy/hotelmanagementapplication/service/RoomService.java [hunk 7]
```diff
@@ -119,6 +160,12 @@ public boolean disableRoom(Long id){
 
     }
 
+    /***
+     * updates a room
+     * @param roomDTO room to be updated
+     * @return the updated room for confirmation
+     * @throws NullPointerException
+     */
```
Comment (DimK10) [inline src/main/java/com/sphy/hotelmanagementapplication/service/RoomService.java:174]: Please, do also return the newly created object, and not the initial one (line 184)

File: src/test/java/com/sphy/hotelmanagementapplication/controller/HotelControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+        // Given
+        HotelDTO hotelDTO = new HotelDTO();
+        hotelDTO.setId(1L);
+        hotelDTO.setName("hotelName");
+
+        // When
+        when(hotelService.getHotelByName(anyString())).thenReturn(hotelDTO);
+
+        // Return
+        mockMvc.perform(get("/api/hotelName/hotelName"))
```
Comment (DimK10) [inline src/test/java/com/sphy/hotelmanagementapplication/controller/HotelControllerTest.java]: Here, you can use the option to add a parameter, like in RoomControllerTest, line 266.

File: src/test/java/com/sphy/hotelmanagementapplication/service/HotelServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+        ClientRepository clientRepository;
+
+        @Mock
+        AdminRepository adminRepository;
+
+        @Mock
+        RoomService roomService;
+
+        @Mock
+        HotelService hotelService;
```
Comment (DimK10) [inline src/test/java/com/sphy/hotelmanagementapplication/service/HotelServiceTest.java:50]: Do not mock the service you test, check the RoomServiceTest - you need to create it with new keyword

File: src/test/java/com/sphy/hotelmanagementapplication/service/HotelServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,412 @@
...
+            hotelDTO1.getRooms().add(roomDTO3);
+
+            RoomDTO roomDTO4 = new RoomDTO();
+            roomDTO4.setId(4L);
+            roomDTO4.setName("roomDTO2");
+            roomDTO4.setHotel(1L);
+            roomDTOS1.add(roomDTO4);
+            hotelDTO.getRooms().add(roomDTO2);
+
+            hotelService = new HotelService(roomRepository, hotelRepository,adminRepository,
```
Comment (DimK10) [inline src/test/java/com/sphy/hotelmanagementapplication/service/HotelServiceTest.java:173]: Since you acctually do it here, where is the need to mock it anyway?

File: src/test/java/com/sphy/hotelmanagementapplication/service/HotelServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+            Optional<Hotel> hotelOptional = Optional.of(hotel);
+
+
+            // when
+            when(hotelRepository.existsById(anyLong())).thenReturn(true);
+            when(hotelRepository.findById(anyLong())).thenReturn(hotelOptional);
+
+
+            //then
+            boolean actual = hotelService.disableHotel(anyLong());
```
Comment (DimK10) [inline src/test/java/com/sphy/hotelmanagementapplication/service/HotelServiceTest.java:380]: anableHotel(anyLong())

**Review by DimK10 — CHANGES_REQUESTED** [review]
Please fix the issues pointed in comments, thanks!

File: src/main/java/com/sphy/hotelmanagementapplication/dto/HotelDTO.java [hunk 12]
```diff
@@ -1,10 +1,12 @@
 package com.sphy.hotelmanagementapplication.dto;
 
+import com.sphy.hotelmanagementapplication.domain.BaseEntity;
+
 import java.io.Serializable;
 import java.util.HashSet;
 import java.util.Set;
 
-public class HotelDTO implements Serializable {
+public class HotelDTO extends BaseEntity implements Serializable{
```
Comment (DimK10) [inline src/main/java/com/sphy/hotelmanagementapplication/dto/HotelDTO.java]: Why did you add BaseEntity?

File: src/main/java/com/sphy/hotelmanagementapplication/dto/HotelDTO.java [hunk 13]
```diff
@@ -21,7 +23,7 @@ public class HotelDTO implements Serializable {
     private Set<RoomDTO> rooms =new HashSet<>();
 
     public HotelDTO(long id) {
-
+        super.setId(id);
```
Comment (DimK10) [inline src/main/java/com/sphy/hotelmanagementapplication/dto/HotelDTO.java]: This is not right, a dto should not extend BaseEntity, we might want to add more functionality to it, and that might break things

File: src/main/java/com/sphy/hotelmanagementapplication/service/RoomService.java [hunk 14]
```diff
@@ -82,9 +84,14 @@ public List<RoomDTO> saveRooms(List<RoomDTO> roomsDTO) throws Exception{
 
         roomsSaved.spliterator().forEachRemaining(rooms::add);
 
+        List<RoomDTO> roomDTOS = new ArrayList<>();
+        for (Room room:rooms){
+            roomDTOS.add(roomToRoomDTO.converter(room));
+        }
+
 		rooms.clear();
```
Comment (DimK10) [inline src/main/java/com/sphy/hotelmanagementapplication/service/RoomService.java]: no need to invoke clear() here, the garbage collector wiil remove everything from memory after this method is executed

File: src/main/java/com/sphy/hotelmanagementapplication/service/RoomService.java [hunk 15]
```diff
@@ -177,11 +184,14 @@ public RoomDTO updateRoom(RoomDTO roomDTO) throws NullPointerException{
...
 			existingRoom.setDisabled(roomDTO.isDisabled());
-            roomRepository.save(existingRoom);
+
             hotel.get().getRooms().add(existingRoom);
             hotelRepository.save(hotel.get());
+
+            return roomToRoomDTO.converter(roomRepository.save(existingRoom));
         }
-        return roomDTO;
+        return null;
```
Comment (DimK10) [inline src/main/java/com/sphy/hotelmanagementapplication/service/RoomService.java:193]: I think, the code will have more quality if we wrapped this in an optional - what do you think? Its is not mandatory to change it if you dont want to.

File: src/test/java/com/sphy/hotelmanagementapplication/controller/HotelControllerTest.java [hunk 16]
```diff
@@ -241,7 +241,7 @@ void findHotelByName() throws Exception {
         when(hotelService.getHotelByName(anyString())).thenReturn(hotelDTO);
 
         // Return
-        mockMvc.perform(get("/api/hotelName/hotelName"))
+        mockMvc.perform(get("/api/hotelName/hotelName", "hotelName"))
```
Comment (DimK10) [inline src/test/java/com/sphy/hotelmanagementapplication/controller/HotelControllerTest.java]: add {} to hotelName inside the string "/api/hotelName/hotelName", and i would personally change the hotelName there to name, so that it is clear to everyone that it is a placeholder value, and not the actual string we want to check (""/api/hotelName/{name}"")

**Review by DimK10 — CHANGES_REQUESTED** [review]
Please check the comments

File: src/main/java/com/sphy/hotelmanagementapplication/service/RoomService.java (hunk 15 above)
Comment (georgepoul) [reply src/main/java/com/sphy/hotelmanagementapplication/service/RoomService.java:193]: it is with optional but if it is not exists it returns null

**Review by DimK10 — APPROVED** [review]
